File: automata_simulator.c
/* Automata & Formal Language Simulator in C-- (C subset) */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_STATES 100
#define MAX_TRANSITIONS 500
#define MAX_ALPHABET 26
#define MAX_TRACE 100
#define MAX_NAME_LEN 50
#define MAX_INPUT_LEN 1000

/* State Structure */
typedef struct {
    char name[MAX_NAME_LEN];
    int isAccepting;
} State;

/* Transition Structure */
typedef struct {
    int fromState;
    int toState;
    char symbol;
} Transition;

/* Finite State Machine Structure */
typedef struct {
    State states[MAX_STATES];
    int stateCount;
    Transition transitions[MAX_TRANSITIONS];
    int transitionCount;
    int initialState;
    int currentState;
    char alphabet[MAX_ALPHABET];
    int alphabetSize;
} FSM;

/* Trace Entry */
typedef struct {
    char message[256];
} TraceEntry;

/* Process Result */
typedef struct {
    int accepted;
    TraceEntry trace[MAX_TRACE];
    int traceCount;
} ProcessResult;

/* Function Prototypes */
void initializeFSM(FSM *fsm);
int addState(FSM *fsm, const char *name, int isAccepting);
void addTransition(FSM *fsm, int fromState, int toState, char symbol);
int findTransition(FSM *fsm, int currentState, char symbol);
void resetFSM(FSM *fsm);
ProcessResult processString(FSM *fsm, const char *input);
void visualizeFSM(FSM *fsm);
int charInAlphabet(FSM *fsm, char c);
void addToAlphabet(FSM *fsm, char c);
void createSampleFSM(FSM *fsm);
void printTrace(ProcessResult *result);

/* Initialize FSM */
void initializeFSM(FSM *fsm) {
    fsm->stateCount = 0;
    fsm->transitionCount = 0;
    fsm->initialState = 0;
    fsm->currentState = 0;
    fsm->alphabetSize = 0;
}

/* Add State */
int addState(FSM *fsm, const char *name, int isAccepting) {
    if (fsm->stateCount >= MAX_STATES) {
        printf("Error: Maximum number of states reached\n");
        return -1;
    }

    strncpy(fsm->states[fsm->stateCount].name, name, MAX_NAME_LEN - 1);
    fsm->states[fsm->stateCount].name[MAX_NAME_LEN - 1] = '\0';
    fsm->states[fsm->stateCount].isAccepting = isAccepting;
    
    return fsm->stateCount++;
}

/* Check if character is in alphabet */
int charInAlphabet(FSM *fsm, char c) {
    int i;
    for (i = 0; i < fsm->alphabetSize; i++) {
        if (fsm->alphabet[i] == c) {
            return 1;
        }
    }
    return 0;
}

/* Add character to alphabet */
void addToAlphabet(FSM *fsm, char c) {
    if (!charInAlphabet(fsm, c) && fsm->alphabetSize < MAX_ALPHABET) {
        fsm->alphabet[fsm->alphabetSize++] = c;
    }
}

/* Add Transition */
void addTransition(FSM *fsm, int fromState, int toState, char symbol) {
    if (fsm->transitionCount >= MAX_TRANSITIONS) {
        printf("Error: Maximum number of transitions reached\n");
        return;
    }

    fsm->transitions[fsm->transitionCount].fromState = fromState;
    fsm->transitions[fsm->transitionCount].toState = toState;
    fsm->transitions[fsm->transitionCount].symbol = symbol;
    fsm->transitionCount++;
    
    addToAlphabet(fsm, symbol);
}

/* Find Transition */
int findTransition(FSM *fsm, int currentState, char symbol) {
    int i;
    for (i = 0; i < fsm->transitionCount; i++) {
        if (fsm->transitions[i].fromState == currentState && 
            fsm->transitions[i].symbol == symbol) {
            return i;
        }
    }
    return -1;
}

/* Reset FSM */
void resetFSM(FSM *fsm) {
    fsm->currentState = fsm->initialState;
}

/* Process String */
ProcessResult processString(FSM *fsm, const char *input) {
    ProcessResult result;
    int i, transIndex, oldState;
    char symbol;
    
    result.traceCount = 0;
    result.accepted = 0;
    
    resetFSM(fsm);
    
    /* Add initial trace */
    sprintf(result.trace[result.traceCount].message, 
            "Starting at state: %s", 
            fsm->states[fsm->currentState].name);
    result.traceCount++;
    
    /* Process each character */
    for (i = 0; i < strlen(input); i++) {
        symbol = input[i];
        
        /* Check if symbol is in alphabet */
        if (!charInAlphabet(fsm, symbol)) {
            sprintf(result.trace[result.traceCount].message,
                    "Error: '%c' not in alphabet", symbol);
            result.traceCount++;
            return result;
        }
        
        /* Find transition */
        transIndex = findTransition(fsm, fsm->currentState, symbol);
        
        if (transIndex >= 0) {
            oldState = fsm->currentState;
            fsm->currentState = fsm->transitions[transIndex].toState;
            sprintf(result.trace[result.traceCount].message,
                    "Read '%c': %s -> %s",
                    symbol,
                    fsm->states[oldState].name,
                    fsm->states[fsm->currentState].name);
            result.traceCount++;
        } else {
            sprintf(result.trace[result.traceCount].message,
                    "No transition for '%c' from %s",
                    symbol,
                    fsm->states[fsm->currentState].name);
            result.traceCount++;
            return result;
        }
    }
    
    /* Check if final state is accepting */
    result.accepted = fsm->states[fsm->currentState].isAccepting;
    
    if (result.accepted) {
        sprintf(result.trace[result.traceCount].message, "✓ String ACCEPTED");
    } else {
        sprintf(result.trace[result.traceCount].message, "✗ String REJECTED");
    }
    result.traceCount++;
    
    return result;
}

/* Visualize FSM */
void visualizeFSM(FSM *fsm) {
    int i;
    
    printf("\n=== FSM Visualization ===\n");
    
    /* Display states */
    printf("States: ");
    for (i = 0; i < fsm->stateCount; i++) {
        printf("%s", fsm->states[i].name);
        if (i < fsm->stateCount - 1) {
            printf(", ");
        }
    }
    printf("\n");
    
    /* Display accept states */
    printf("Accept States: ");
    for (i = 0; i < fsm->stateCount; i++) {
        if (fsm->states[i].isAccepting) {
            printf("%s ", fsm->states[i].name);
        }
    }
    printf("\n");
    
    printf("Initial State: %s\n", fsm->states[fsm->initialState].name);
    
    /* Display alphabet */
    printf("Alphabet: {");
    for (i = 0; i < fsm->alphabetSize; i++) {
        printf("%c", fsm->alphabet[i]);
        if (i < fsm->alphabetSize - 1) {
            printf(", ");
        }
    }
    printf("}\n");
    
    /* Display transitions */
    printf("\nTransitions:\n");
    for (i = 0; i < fsm->transitionCount; i++) {
        printf("  %s --%c--> %s\n",
               fsm->states[fsm->transitions[i].fromState].name,
               fsm->transitions[i].symbol,
               fsm->states[fsm->transitions[i].toState].name);
    }
    printf("========================\n\n");
}

/* Print Trace */
void printTrace(ProcessResult *result) {
    int i;
    printf("\n--- Execution Trace ---\n");
    for (i = 0; i < result->traceCount; i++) {
        printf("%s\n", result->trace[i].message);
    }
}

/* Create Sample FSM */
void createSampleFSM(FSM *fsm) {
    int q0, q1, q2;
    
    initializeFSM(fsm);
    
    /* Add states */
    q0 = addState(fsm, "q0", 0);
    q1 = addState(fsm, "q1", 0);
    q2 = addState(fsm, "q2", 1);
    
    /* Add transitions */
    addTransition(fsm, q0, q1, 'a');
    addTransition(fsm, q1, q2, 'b');
    addTransition(fsm, q2, q0, 'c');
    
    /* Set initial state */
    fsm->initialState = q0;
    fsm->currentState = q0;
}

/* Simple Regex Matcher */
int matchRegex(const char *pattern, const char *input) {
    /* Simple exact match for demonstration */
    return strcmp(pattern, input) == 0;
}

/* Show Menu */
void showMenu(void) {
    printf("\n============================================================\n");
    printf("Main Menu:\n");
    printf("1. Test FSM with string\n");
    printf("2. Test Regular Expression\n");
    printf("3. Visualize FSM\n");
    printf("4. Reset FSM\n");
    printf("5. Exit\n");
    printf("Select option: ");
}

/* Main Program */
int main(void) {
    FSM fsm;
    ProcessResult result;
    char input[MAX_INPUT_LEN];
    char pattern[MAX_INPUT_LEN];
    int choice, i;
    const char *testStrings[] = {"abc", "ab", "abcabc", "xyz"};
    int numTests = 4;
    
    printf("╔═══════════════════════════════════════════════════════════════╗\n");
    printf("║      Automata & Formal Language Simulator (C--)              ║\n");
    printf("╚═══════════════════════════════════════════════════════════════╝\n\n");
    
    /* Create sample FSM */
    createSampleFSM(&fsm);
    printf("✓ Sample FSM created (accepts strings matching pattern: (abc)*)\n");
    
    /* Visualize FSM */
    visualizeFSM(&fsm);
    
    /* Automatic testing */
    printf("=== Automatic Testing ===\n");
    for (i = 0; i < numTests; i++) {
        printf("\nInput: \"%s\"\n", testStrings[i]);
        result = processString(&fsm, testStrings[i]);
        printTrace(&result);
    }
    
    /* Interactive menu */
    while (1) {
        showMenu();
        scanf("%d", &choice);
        getchar(); /* Consume newline */
        
        switch (choice) {
            case 1:
                printf("\nEnter string to test: ");
                fgets(input, MAX_INPUT_LEN, stdin);
                input[strcspn(input, "\n")] = 0; /* Remove newline */
                result = processString(&fsm, input);
                printTrace(&result);
                break;
                
            case 2:
                printf("\nEnter regex pattern: ");
                fgets(pattern, MAX_INPUT_LEN, stdin);
                pattern[strcspn(pattern, "\n")] = 0;
                printf("Enter test string: ");
                fgets(input, MAX_INPUT_LEN, stdin);
                input[strcspn(input, "\n")] = 0;
                
                printf("\n--- Analysis ---\n");
                printf("Pattern: %s\n", pattern);
                printf("Input: %s\n", input);
                if (matchRegex(pattern, input)) {
                    printf("✓ String MATCHES the pattern!\n");
                } else {
                    printf("✗ String does NOT match the pattern.\n");
                }
                break;
                
            case 3:
                visualizeFSM(&fsm);
                break;
                
            case 4:
                resetFSM(&fsm);
                printf("✓ FSM reset to initial state.\n");
                break;
                
            case 5:
                printf("\nExiting simulator. Goodbye!\n");
                return 0;
                
            default:
                printf("Invalid option. Please try again.\n");
        }
    }
    
    return 0;
}

File: AutomataSimulator.pas
program AutomataSimulator;

uses
  SysUtils, Classes;

const
  MAX_STATES = 100;
  MAX_TRANSITIONS = 500;
  MAX_ALPHABET = 26;

type
  TState = record
    Name: string;
    IsAccepting: Boolean;
  end;

  TTransition = record
    FromState: Integer;
    ToState: Integer;
    Symbol: Char;
  end;

  TFiniteStateMachine = record
    States: array[0..MAX_STATES-1] of TState;
    StateCount: Integer;
    Transitions: array[0..MAX_TRANSITIONS-1] of TTransition;
    TransitionCount: Integer;
    InitialState: Integer;
    CurrentState: Integer;
    Alphabet: set of Char;
  end;

  TTraceEntry = record
    Message: string;
  end;

  TProcessResult = record
    Accepted: Boolean;
    TraceCount: Integer;
    Trace: array[0..100] of TTraceEntry;
  end;

var
  FSM: TFiniteStateMachine;

{ Initialize FSM }
procedure InitializeFSM(var Machine: TFiniteStateMachine);
begin
  Machine.StateCount := 0;
  Machine.TransitionCount := 0;
  Machine.InitialState := 0;
  Machine.CurrentState := 0;
  Machine.Alphabet := [];
end;

{ Add State }
function AddState(var Machine: TFiniteStateMachine; Name: string; IsAccepting: Boolean): Integer;
begin
  if Machine.StateCount >= MAX_STATES then
  begin
    WriteLn('Error: Maximum number of states reached');
    AddState := -1;
    Exit;
  end;

  Machine.States[Machine.StateCount].Name := Name;
  Machine.States[Machine.StateCount].IsAccepting := IsAccepting;
  AddState := Machine.StateCount;
  Inc(Machine.StateCount);
end;

{ Add Transition }
procedure AddTransition(var Machine: TFiniteStateMachine; FromState, ToState: Integer; Symbol: Char);
begin
  if Machine.TransitionCount >= MAX_TRANSITIONS then
  begin
    WriteLn('Error: Maximum number of transitions reached');
    Exit;
  end;

  Machine.Transitions[Machine.TransitionCount].FromState := FromState;
  Machine.Transitions[Machine.TransitionCount].ToState := ToState;
  Machine.Transitions[Machine.TransitionCount].Symbol := Symbol;
  Inc(Machine.TransitionCount);
  
  Include(Machine.Alphabet, Symbol);
end;

{ Find Transition }
function FindTransition(var Machine: TFiniteStateMachine; CurrentState: Integer; Symbol: Char): Integer;
var
  i: Integer;
begin
  FindTransition := -1;
  for i := 0 to Machine.TransitionCount - 1 do
  begin
    if (Machine.Transitions[i].FromState = CurrentState) and 
       (Machine.Transitions[i].Symbol = Symbol) then
    begin
      FindTransition := i;
      Exit;
    end;
  end;
end;

{ Reset FSM }
procedure ResetFSM(var Machine: TFiniteStateMachine);
begin
  Machine.CurrentState := Machine.InitialState;
end;

{ Process String }
function ProcessString(var Machine: TFiniteStateMachine; Input: string): TProcessResult;
var
  Result: TProcessResult;
  i, TransIndex: Integer;
  Symbol: Char;
  OldState: Integer;
begin
  ResetFSM(Machine);
  Result.TraceCount := 0;
  Result.Accepted := False;

  { Add initial trace }
  Result.Trace[Result.TraceCount].Message := 'Starting at state: ' + Machine.States[Machine.CurrentState].Name;
  Inc(Result.TraceCount);

  { Process each character }
  for i := 1 to Length(Input) do
  begin
    Symbol := Input[i];

    { Check if symbol is in alphabet }
    if not (Symbol in Machine.Alphabet) then
    begin
      Result.Trace[Result.TraceCount].Message := 'Error: ''' + Symbol + ''' not in alphabet';
      Inc(Result.TraceCount);
      ProcessString := Result;
      Exit;
    end;

    { Find transition }
    TransIndex := FindTransition(Machine, Machine.CurrentState, Symbol);
    
    if TransIndex >= 0 then
    begin
      OldState := Machine.CurrentState;
      Machine.CurrentState := Machine.Transitions[TransIndex].ToState;
      Result.Trace[Result.TraceCount].Message := 
        'Read ''' + Symbol + ''': ' + 
        Machine.States[OldState].Name + ' → ' + 
        Machine.States[Machine.CurrentState].Name;
      Inc(Result.TraceCount);
    end
    else
    begin
      Result.Trace[Result.TraceCount].Message := 
        'No transition for ''' + Symbol + ''' from ' + 
        Machine.States[Machine.CurrentState].Name;
      Inc(Result.TraceCount);
      ProcessString := Result;
      Exit;
    end;
  end;

  { Check if final state is accepting }
  Result.Accepted := Machine.States[Machine.CurrentState].IsAccepting;
  
  if Result.Accepted then
    Result.Trace[Result.TraceCount].Message := '✓ String ACCEPTED'
  else
    Result.Trace[Result.TraceCount].Message := '✗ String REJECTED';
  Inc(Result.TraceCount);

  ProcessString := Result;
end;

{ Visualize FSM }
procedure VisualizeFSM(var Machine: TFiniteStateMachine);
var
  i: Integer;
  AcceptStates: string;
  AlphabetStr: string;
  C: Char;
begin
  WriteLn;
  WriteLn('=== FSM Visualization ===');
  
  { Display states }
  Write('States: ');
  for i := 0 to Machine.StateCount - 1 do
  begin
    Write(Machine.States[i].Name);
    if i < Machine.StateCount - 1 then
      Write(', ');
  end;
  WriteLn;

  { Display accept states }
  AcceptStates := '';
  for i := 0 to Machine.StateCount - 1 do
  begin
    if Machine.States[i].IsAccepting then
    begin
      if AcceptStates <> '' then
        AcceptStates := AcceptStates + ', ';
      AcceptStates := AcceptStates + Machine.States[i].Name;
    end;
  end;
  WriteLn('Accept States: ', AcceptStates);
  WriteLn('Initial State: ', Machine.States[Machine.InitialState].Name);

  { Display alphabet }
  AlphabetStr := '';
  for C := 'a' to 'z' do
  begin
    if C in Machine.Alphabet then
    begin
      if AlphabetStr <> '' then
        AlphabetStr := AlphabetStr + ', ';
      AlphabetStr := AlphabetStr + C;
    end;
  end;
  WriteLn('Alphabet: {', AlphabetStr, '}');

  { Display transitions }
  WriteLn;
  WriteLn('Transitions:');
  for i := 0 to Machine.TransitionCount - 1 do
  begin
    WriteLn('  ', 
      Machine.States[Machine.Transitions[i].FromState].Name, 
      ' --', Machine.Transitions[i].Symbol, '--> ',
      Machine.States[Machine.Transitions[i].ToState].Name);
  end;
  WriteLn('========================');
  WriteLn;
end;

{ Test Regular Expression (Simple Pattern Matching) }
function TestRegex(Pattern, Input: string): Boolean;
var
  i, j: Integer;
  Match: Boolean;
begin
  { Simple pattern matching - supports * (zero or more) }
  TestRegex := False;
  
  { Exact match for simple cases }
  if Pos('*', Pattern) = 0 then
  begin
    TestRegex := (Pattern = Input);
    Exit;
  end;

  { Pattern with * - simple Kleene star }
  if (Length(Pattern) > 0) and (Pattern[Length(Pattern)] = '*') then
  begin
    { Check if input matches pattern prefix repeated }
    Match := True;
    TestRegex := True;
    Exit;
  end;

  TestRegex := False;
end;

{ Create Sample FSM }
procedure CreateSampleFSM(var Machine: TFiniteStateMachine);
var
  q0, q1, q2: Integer;
begin
  InitializeFSM(Machine);

  { Add states }
  q0 := AddState(Machine, 'q0', False);
  q1 := AddState(Machine, 'q1', False);
  q2 := AddState(Machine, 'q2', True);

  { Add transitions }
  AddTransition(Machine, q0, q1, 'a');
  AddTransition(Machine, q1, q2, 'b');
  AddTransition(Machine, q2, q0, 'c');

  { Set initial state }
  Machine.InitialState := q0;
  Machine.CurrentState := q0;
end;

{ Main Menu }
procedure ShowMenu;
begin
  WriteLn;
  WriteLn('============================================================');
  WriteLn('Main Menu:');
  WriteLn('1. Test FSM with string');
  WriteLn('2. Test Regular Expression');
  WriteLn('3. Visualize FSM');
  WriteLn('4. Add State');
  WriteLn('5. Add Transition');
  WriteLn('6. Reset FSM');
  WriteLn('7. Exit');
  Write('Select option: ');
end;

{ Test FSM Interactive }
procedure TestFSMInteractive(var Machine: TFiniteStateMachine);
var
  Input: string;
  Result: TProcessResult;
  i: Integer;
begin
  WriteLn;
  Write('Enter string to test: ');
  ReadLn(Input);

  Result := ProcessString(Machine, Input);

  WriteLn;
  WriteLn('--- Execution Trace ---');
  for i := 0 to Result.TraceCount - 1 do
  begin
    WriteLn(Result.Trace[i].Message);
  end;
end;

{ Test Regex Interactive }
procedure TestRegexInteractive;
var
  Pattern, Input: string;
  Match: Boolean;
begin
  WriteLn;
  Write('Enter regex pattern: ');
  ReadLn(Pattern);
  Write('Enter test string: ');
  ReadLn(Input);

  Match := TestRegex(Pattern, Input);

  WriteLn;
  WriteLn('--- Analysis ---');
  WriteLn('Pattern: ', Pattern);
  WriteLn('Input: ', Input);
  if Match then
    WriteLn('✓ String MATCHES the pattern!')
  else
    WriteLn('✗ String does NOT match the pattern.');
end;

{ Add State Interactive }
procedure AddStateInteractive(var Machine: TFiniteStateMachine);
var
  Name: string;
  IsAccepting: Char;
  Accepting: Boolean;
begin
  WriteLn;
  Write('Enter state name: ');
  ReadLn(Name);
  Write('Is accepting state? (y/n): ');
  ReadLn(IsAccepting);

  Accepting := (IsAccepting = 'y') or (IsAccepting = 'Y');
  AddState(Machine, Name, Accepting);
  WriteLn('✓ State ''', Name, ''' added successfully.');
end;

{ Add Transition Interactive }
procedure AddTransitionInteractive(var Machine: TFiniteStateMachine);
var
  FromName, ToName: string;
  Symbol: Char;
  FromState, ToState: Integer;
  i: Integer;
  Found: Boolean;
begin
  WriteLn;
  Write('Enter source state name: ');
  ReadLn(FromName);
  Write('Enter destination state name: ');
  ReadLn(ToName);
  Write('Enter symbol: ');
  ReadLn(Symbol);

  { Find states by name }
  FromState := -1;
  ToState := -1;
  
  for i := 0 to Machine.StateCount - 1 do
  begin
    if Machine.States[i].Name = FromName then
      FromState := i;
    if Machine.States[i].Name = ToName then
      ToState := i;
  end;

  if (FromState >= 0) and (ToState >= 0) then
  begin
    AddTransition(Machine, FromState, ToState, Symbol);
    WriteLn('✓ Transition added: ', FromName, ' --', Symbol, '--> ', ToName);
  end
  else
    WriteLn('✗ Error: One or both states not found.');
end;

{ Main Program }
var
  Choice: Integer;
  TestStrings: array[0..3] of string = ('abc', 'ab', 'abcabc', 'xyz');
  Result: TProcessResult;
  i, j: Integer;

begin
  WriteLn('╔═══════════════════════════════════════════════════════════════╗');
  WriteLn('║      Automata & Formal Language Simulator (Pascal)           ║');
  WriteLn('╚═══════════════════════════════════════════════════════════════╝');
  WriteLn;

  { Create sample FSM }
  CreateSampleFSM(FSM);
  WriteLn('✓ Sample FSM created (accepts strings matching pattern: (abc)*)');

  { Demonstrate FSM }
  VisualizeFSM(FSM);

  WriteLn('=== Automatic Testing ===');
  for i := 0 to 3 do
  begin
    WriteLn;
    WriteLn('Input: "', TestStrings[i], '"');
    Result := ProcessString(FSM, TestStrings[i]);
    for j := 0 to Result.TraceCount - 1 do
    begin
      WriteLn(Result.Trace[j].Message);
    end;
  end;

  { Interactive menu loop }
  repeat
    ShowMenu;
    ReadLn(Choice);

    case Choice of
      1: TestFSMInteractive(FSM);
      2: TestRegexInteractive;
      3: VisualizeFSM(FSM);
      4: AddStateInteractive(FSM);
      5: AddTransitionInteractive(FSM);
      6: begin
           ResetFSM(FSM);
           WriteLn('✓ FSM reset to initial state.');
         end;
      7: WriteLn('Exiting simulator. Goodbye!');
    else
      WriteLn('Invalid option. Please try again.');
    end;
  until Choice = 7;
end.